File: app/src/test/java/com/dougkeen/bart/model/StationPairTest.java
package com.dougkeen.bart.model;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.junit.Test;

public class StationPairTest {
    @Test
    public void fareUpdatesCreateAReplacementValue() {
        StationPair original = new StationPair(Station.MONT, Station.RICH);

        StationPair updated = original.withFare("$5.00", 123L);

        assertNull(original.getFare());
        assertEquals(0L, original.getFareLastUpdated());
        assertEquals("$5.00", updated.getFare());
        assertEquals(123L, updated.getFareLastUpdated());
        assertEquals(original.getOrigin(), updated.getOrigin());
        assertEquals(original.getDestination(), updated.getDestination());
    }

    @Test
    public void jsonRoundTripPreservesFareMetadata() throws Exception {
        StationPair original = new StationPair(Station.MONT, Station.RICH,
                "$5.00", 123L, 42, 7);
        ObjectMapper mapper = new ObjectMapper()
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES,
                        false);

        StationPair restored = mapper.readValue(mapper.writeValueAsString(original),
                StationPair.class);

        assertEquals(original.getOrigin(), restored.getOrigin());
        assertEquals(original.getDestination(), restored.getDestination());
        assertEquals(original.getFare(), restored.getFare());
        assertEquals(original.getFareLastUpdated(), restored.getFareLastUpdated());
        assertEquals(original.getAverageTripLength(), restored.getAverageTripLength());
        assertEquals(original.getAverageTripSampleCount(),
                restored.getAverageTripSampleCount());
    }
}

File: app/src/test/kotlin/com/dougkeen/bart/model/StationPairTest.kt
package com.dougkeen.bart.model

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class StationPairTest {
    @Test
    fun fareUpdatesCreateAReplacementValue() {
        val original = StationPair(Station.MONT, Station.RICH)
        val updated = original.withFare("$5.00", 123L)
        assertNull(original.fare)
        assertEquals(0L, original.fareLastUpdated)
        assertEquals("$5.00", updated.fare)
        assertEquals(123L, updated.fareLastUpdated)
        assertEquals(original.origin, updated.origin)
        assertEquals(original.destination, updated.destination)
    }

    @Test
    fun jsonRoundTripPreservesFareMetadata() {
        val original = StationPair(Station.MONT, Station.RICH, "$5.00", 123L, 42, 7)
        val mapper = ObjectMapper().configure(
            DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES,
            false,
        )
        val restored = mapper.readValue(mapper.writeValueAsString(original), StationPair::class.java)
        assertEquals(original.origin, restored.origin)
        assertEquals(original.destination, restored.destination)
        assertEquals(original.fare, restored.fare)
        assertEquals(original.fareLastUpdated, restored.fareLastUpdated)
        assertEquals(original.averageTripLength, restored.averageTripLength)
        assertEquals(original.averageTripSampleCount, restored.averageTripSampleCount)
    }
}
